deeplearning4j-nearestneighbors-parent/nearestneighbor-core/src/main/java/org/deeplearning4j/clustering/lsh/RandomProjectionLSH.java

```diff
@@ -0,0 +1,230 @@
+package org.deeplearning4j.clustering.lsh;
+
+import lombok.Getter;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.api.ops.impl.broadcast.BroadcastDivOp;
+import org.nd4j.linalg.api.ops.impl.broadcast.BroadcastEqualTo;
+import org.nd4j.linalg.api.ops.impl.transforms.Sign;
+
+import org.nd4j.linalg.api.ops.random.impl.GaussianDistribution;
+import org.nd4j.linalg.api.ops.random.impl.UniformDistribution;
+import org.nd4j.linalg.api.rng.Random;
+import org.nd4j.linalg.exception.ND4JIllegalStateException;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.BooleanIndexing;
+import org.nd4j.linalg.indexing.conditions.Conditions;
+import org.nd4j.linalg.ops.transforms.Transforms;
+
+import java.util.Arrays;
+
+
+/**
+ * This class implements Entropy LSH for the cosine distance, in order to preserve memory for large datasets.
+ *
+ * Entropy SLH is the LSH scheme of
+ *
+ * _Entropy based nearest neighbor search in high dimensions_
+ * R Panigrahy - SIAM 2006
+ * https://arxiv.org/pdf/cs/0510019.pdf
+ *
+ * To read more about LSH, in particular for the Cosine distance, see
+ * chapter 3 of :
+ * _Mining Massive Datasets_, Anand Rajaraman and Jeffrey Ullman
+ * http://www.mmds.org/
+ *
+ * The original development of LSH for the cosine distance is from
+ * Similarity estimation techniques from rounding algorithms
+ * MS Charikar - STOCS, 2002
+ *
+ * Note for high-precision or distributed settings, you should not
+ * use this and rather extend this to layered LSH ( https://arxiv.org/abs/1210.7057 )
+ *
+ */
+public class RandomProjectionLSH implements LSH {
+
+    @Override
+    public String getDistanceMeasure(){
+        return "cosinedistance";
+    }
+
+    @Getter int hashLength;
+
+    @Getter int numTables;
+
+    @Getter int inDimension;
+
+
+    @Getter double radius;
+
+    INDArray randomProjection;
+
+    INDArray index;
+
+    INDArray indexData;
+
+
+    private INDArray gaussianRandomMatrix(int[] shape, Random rng){
+        INDArray res = Nd4j.create(shape);
+
+        GaussianDistribution op1 = new GaussianDistribution(res, 0.0, 1.0 / Math.sqrt(shape[0]));
+
+        Nd4j.getExecutioner().exec(op1, rng);
+        return res;
+    }
+
+    public RandomProjectionLSH(int hashLength, int numTables, int inDimension, double radius){
+        this(hashLength, numTables, inDimension, radius, Nd4j.getRandom());
+    }
```

Comment:
Builder maybe?

Reply:
I'm not sure: `hashLength`, `numTables`, `inDimension`, `radius`, are all required and not one has a sensible default.